### ISSUE TITLE
Cannot process Porcupine in Pixel 4a [android 11] when app targetSdkVersion is 30

### ISSUE THREAD

**Issue by orcuns** [Picovoice/porcupine#307]
I have configured [Service based sample app](https://github.com/Picovoice/porcupine/tree/master/demo/android/Service). And it works quite good. 

I have a Pixel 4a and I wanted to set `targetSdkVersion to 30` and `compileSdkVersion 30`. 

But unfortunately I encounter with this: 
`com.example.androidforegroundservice I/PICOVOICE: [ERROR] invalid object cannot be used for processing.`

It works as expected when I downgrade `targetSdkVersion to 29`

### set app build.gradle file:

```
android {
    compileSdkVersion 30
    buildToolsVersion "30.0.2"
    defaultConfig {
        applicationId "com.example.androidforegroundservice"
        minSdkVersion 15
        targetSdkVersion 30 -> I want to set it to latest targetSdkVersion and Porcupine can process
       .....
    }
......
}
```

## Actual behaviour

I receive the the error that I wrote above. You can check the screenshot below:

![Screenshot 2020-10-16 at 10 47 55](https://user-images.githubusercontent.com/5539830/96237385-ed49f680-0fa5-11eb-80f9-c6739a6bc10c.png)

## Steps to reproduce the behaviour

You can increase `targetSdkVersion to 30` in order to reproduce the behaviour.

**Comment by kenarsa** [comment]
The latest version is setting it to 30. Have you tried that?

**Comment by orcuns** [comment]
> The latest version is setting it to 30. Have you tried that?

Yes, when you set it to 30, Porcupine starts to give an error. If you downgrade to 29, it works!

**Comment by kenarsa** [comment]
I think you misunderstood. The latest code on GitHub is setting it to 30! do a git pull, clean your build, and try it again, please.

**Comment by orcuns** [comment]
> I think you misunderstood. The latest code on GitHub is setting it to 30! do a git pull, clean your build, and try it again, please.

Okay, I pulled the latest Porcupine code from master. Cleaned my build. Now I can't see the error I mentioned before; but Wakeword [Porcupine, Bumblebee] doesn't react. If I downgrade to 29, Wakeword's caught and notification counter rises up. 

I assume, it is a device [Pixel 4a] based problem.

**Comment by kenarsa** [comment]
Not sure why it doesn't trigger. I doubt it is a phone problem though. Can you try to download the same app below?

https://play.google.com/store/apps/details?id=ai.picovoice.porcupine.demo&hl=en_CA

Let me know if you can run this. If you can and it works.

**Comment by orcuns** [comment]
Unfortunately it doesn't work with demo app too.

I also tested with OnePlus 7 Pro (android 10), Nexus 6P (android 8.1) and Pixel 3(android 11). Works with them.

I think Porcupine has a problem with Pixel 4a. If you have a Pixel 4a, could you also verify the issue?

**Comment by kenarsa** [comment]
I believe we did try it on Android 11 and Pixel 3/2 and it worked just fine. I hardly think its a Pixel issue. I think something is off in the settings on the phone you are working with. Can you do a factory reset and make sure you are giving microphone permission?

If that doesn't work I suggest looking into logcat and see if there is an obvious error message there.

As the last step, you can record/dump the audio passed to Porcupine. Simply concatenate buffers passed to engine here https://github.com/Picovoice/porcupine/blob/master/binding/android/Porcupine/porcupine/src/main/java/ai/picovoice/porcupine/PorcupineManager.java#L93 and then dump to a file. In order to do this, you need to rebuild the binding from source and update its jar in the demo.

**Comment by orcuns** [comment]
- Microphone permission is given. I think it's not the issue here.

- I tried factory reset, nothing changed.

- As the last step, I tried with Om Recorder to record and passed audio to Porcupine like you said. I still see the same error.

`PORCUPINE ERROR: ai.picovoice.porcupine.PorcupineException: java.lang.IllegalArgumentException: Invalid object ID.`


Anyway, I think I won't target the app to 30 until it's fixed. But I seriously suggest that you should check on Pixel 4a.

**Comment by kenarsa** [comment]
Sounds good. I keep the issue open to see if other users report the same. AVD does not provide respective emulator otherwise we could try to reproduce there.

**Comment by nanaghartey** [comment]
I'm also facing the same issue. The android demo app and the version on the play store isn't working on my pixel 4 xl running android 11. I tried on older devices and it worked but for the pixel 4xl, it fails

**Comment by kenarsa** [comment]
@nanaghartey I tried the demo as is on Pixel 4 XL emulator and it works as expected. I guess we need to run it on a physical device to reproduce. Can you try the demo on the emulator and see if you can reproduce?

**Comment by nanaghartey** [comment]
@kenarsa I can confirm that:
-Sample app works on the pixel 4 XL physical device when `targetSdkVersion` is set to 29. It fails when set to 30. Perhaps a file access issue on Android 11
- Demo app on play store fails on the pixel 4 XL device
Unfortunately, I'm unable to test on an emulator; only on a physical device

**Comment by pborsutzki** [comment]
I am facing the same issue with a Pixel 4a on Android 11. I inserted a breakpoint into the catch clause in [Porcupine.process](https://github.com/Picovoice/porcupine/blob/df29f2447884b21995265241a2774f37bb751532/binding/android/Porcupine/porcupine/src/main/java/ai/picovoice/porcupine/Porcupine.java#L90) and found that in the first call to this function an exception is being thrown: `java.lang.IllegalArgumentException: Invalid object ID.`
`this.handle` is set to something nonzero and `pcm` is an array of 512 shorts.

Setting `targetSdkVersion` to 29 works around the issue.

**Comment by johnboy00** [comment]
Same issue on Pixel 5 running Android 11.  Play Store demo app doesn't work, and my own test app doesn't work ("invalid object ID") when targeting SDK 30.

**Comment by kenarsa** [comment]
@johnboy00 can you reproduce this on an emulator as well or is it just on the physical device?

**Comment by johnboy00** [comment]
> @johnboy00 can you reproduce this on an emulator as well or is it just on the physical device?

I have not tried to reproduce on an emulator.  I've only tried on my personal device.  Pixel 5 + Android 11 + March 5, 2021 security update.

**Comment by hannesa2** [comment]
I can confirm 

- Pixel 4 with Android 11, targetSdk 30 doesn't work
- Pixel 2 with Android 11, targetSdk 30 works
- [Apple Silicon Emulator Android 11](https://github.com/741g/android-emulator-m1-preview/issues/10#issuecomment-773532762), targetSdk 30 doesn't work
- [Apple Silicon Emulator Android 11](https://github.com/741g/android-emulator-m1-preview/issues/10#issuecomment-773532762), targetSdk 29 works

They changed something in Emulator audio https://issuetracker.google.com/issues/152811102#comment7 and https://issuetracker.google.com/issues/157223337 That's why only the most recent Emulator will probably work on macOS

Does someone knows the root cause why only targetSdk 29 works on Pixel 4 ?

**Comment by laves** [comment]
We've just merged in a fix for this (#409). We can confirm that this fix addresses the issue on a physical Pixel 4a device targeting SDK 30.

**Comment by laves** [comment]
If you're using the maven package, update to porcupine-android:1.9.1 to get the fix.
